gridfs/grid_file.py

```diff
@@ -179,7 +179,7 @@ def __create_index(self, collection, index, unique):
                 indexes = list(collection.list_indexes())
             except OperationFailure:
                 indexes = []
-            if index not in indexes:
+            if dict(index) not in [idx['key'].to_dict() for idx in indexes]:
```

Comment:
Actually, this code is more wrong than we thought. `list_indexes()` returns the full index spec but we're trying to compare that to just the index keys. Instead we should look at only the `'key'` field in the index spec:
```python
try:
    index_keys =[index_spec['key'] for index_spec in collection.list_indexes()]
except OperationFailure:
    index_keys = []
if index_key not in index_keys:
...
```

On your point about comment ordering: `list_indexes()` returns the indexes as a SON document so the keys will be well ordered. The order of keys in indexes are important; the index `SON([('filename', 1), ('uploadDate', 1)])` is different from `SON([('uploadDate', 1), ('filename', 1)])`. That's why we should not convert these to dictionaries.